kubemarine/packages.py

```diff
@@ -12,64 +12,191 @@
-                            os_specific_associations[association]['package_name'].split('={{')[0]
-            else:
-                raise Exception('Unexpected value for association')
+    os_family = different_os[0][0]
+    if os_family in ('unknown', 'unsupported'):
+        # For add_node/install procedures we check that OS is supported in prepare.check.system task.
+        # For check_iaas procedure it is allowed to have unsupported OS, so skip caching.
+        cluster.log.debug("Skip caching of packages for unsupported OS.")
+        return inventory
 
-    else:
-        # set 'skip_caching' for customer association
-        if cluster.raw_inventory.get('services', {}).get('packages', {}).get('associations', {}):
-            for package in cluster.raw_inventory['services']['packages']['associations']:
-                os_specific_associations[package]['skip_caching'] = "true"
+    nodes_cache_versions = cluster.nodes['all'].get_final_nodes().get_sudo_nodes()
+    if nodes_cache_versions.is_empty():
+        # For add_node/install procedures we check that all nodes are sudoers in prepare.check.sudoer task.
+        # For check_iaas procedure the nodes might still be not sudoers, so skip caching.
+        cluster.log.debug(f"There are no nodes with sudo privileges, packages will not be cached.")
+        return inventory
+
+    packages_list = _get_packages_to_detect_versions(cluster, inventory, ensured_associations_only)
+    detected_packages = detect_installed_packages_version_groups(nodes_cache_versions, packages_list)
+
+    _cache_package_associations(cluster, inventory, detected_packages, ensured_associations_only)
+    _cache_custom_packages(cluster, inventory, detected_packages, ensured_associations_only)
+
+    cluster.log.debug('Package versions detection finished')
+    return inventory
+
+
+def _get_associations(cluster: KubernetesCluster, inventory: dict):
+    return inventory['services']['packages']['associations'][cluster.get_os_family()]
+
+
+def _get_package_names_for_association(cluster: KubernetesCluster, inventory: dict, association_name: str) -> list:
+    if association_name in get_associations_os_family_keys():
+        return []
+
+    associated_packages = _get_associations(cluster, inventory)[association_name].get('package_name')
+    if isinstance(associated_packages, str):
+        associated_packages = [associated_packages]
+    elif not isinstance(associated_packages, list):
+        raise Exception('Unsupported associated packages object type')
+
+    return associated_packages
+
+
+def _get_packages_for_associations_to_detect(cluster: KubernetesCluster, inventory: dict, association_name: str,
+                                             ensured_association_only: bool) -> list:
+    packages_list = _get_package_names_for_association(cluster, inventory, association_name)
+    if not packages_list:
+        return []
+
+    global_cache_versions = inventory['services']['packages']['cache_versions']
```

Comment:
I suggest to rename the option to `ensure_same_versions` as it now works only for `add_node` and can be thought of as we guarantee same versions on nodes. What do you think?

Reply:
let's leave as is